### ISSUE TITLE
ASSERT FAILED at /pytorch/aten/src/ATen/core/jit_type.h:128 [bug]

### ISSUE THREAD

**Issue by Magnetar99** [pytorch/pytorch#18903]
If you have a question or would like help and support, please ask at our
[forums](https://discuss.pytorch.org/).

If you are submitting a feature request, please preface the title with [feature request].
If you are submitting a bug report, please fill in the following details.

## Issue description

Provide a short description.

## Code example
```
RuntimeError: r ASSERT FAILED at /pytorch/aten/src/ATen/core/jit_type.h:128, please report a bug to PyTorch. (expect at /pytorch/aten/src/ATen/core/jit_type.h:128)
frame #0: std::function<std::string ()>::operator()() const + 0x11 (0x7f471cec0021 in /opt/conda/lib/python3.6/site-packages/torch/lib/libc10.so)
frame #1: c10::Error::Error(c10::SourceLocation, std::string const&) + 0x2a (0x7f471cebf8ea in /opt/conda/lib/python3.6/site-packages/torch/lib/libc10.so)
frame #2: std::shared_ptr<c10::CompleteTensorType> c10::Type::expect<c10::CompleteTensorType>() + 0xa3 (0x7f475744ef93 in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
frame #3: <unknown function> + 0x3793cb (0x7f47574653cb in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
frame #4: <unknown function> + 0x11642e (0x7f475720242e in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
<omitting python frames>
frame #41: __libc_start_main + 0xe7 (0x7f475d5efb97 in /lib/x86_64-linux-gnu/libc.so.6)
```

Please try to provide a minimal example to repro the bug.
Error messages and stack traces are also helpful.

## System Info
Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).
PyTorch version: 1.0.1.post2
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 18.04.2 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: version 3.13.3

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 410.78
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.5.0

Versions of relevant libraries:
[pip] numpy==1.16.2
[pip] torch==1.0.1.post2
[pip] torchvision==0.2.2.post3
[conda] nomkl                     3.0                           0    defaults
[conda] torch                     1.0.1.post2              pypi_0    pypi
[conda] torchvision               0.2.2.post3              pypi_0    pypi
You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

- PyTorch or Caffe2:
- How you installed PyTorch (conda, pip, source):
- Build command you used (if compiling from source):
- OS:
- PyTorch version:
- Python version:
- CUDA/cuDNN version:
- GPU models and configuration:
- GCC version (if compiling from source):
- CMake version:
- Versions of any other relevant libraries:

**Comment by suo** [comment]
Thanks for the report! Could you provide a small script we could run to reproduce the problem? Otherwise it is difficult for us to investigate.

**Comment by SantoshManicka** [comment]
I ran into a very similar, if not the same, error. Below is the error log. I'm not sure I could reproduce it, since it seems rare (I've seen it only once). This is on version 1.0.1 on RHEL.

terminate called after throwing an instance of 'c10::Error'
  what():  owning_ptr == NullType::singleton() || owning_ptr->refcount_.load() > 0 ASSERT FAILED at /opt/conda/conda-bld/pytorch_1549636813070/work/c10/util/intrusive_ptr.h:350, please report a bug to PyTorch. intrusive_ptr: Can only intrusive_ptr::reclaim() owning pointers that were created using intrusive_ptr::release(). (reclaim at /opt/conda/conda-bld/pytorch_1549636813070/work/c10/util/intrusive_ptr.h:350)
frame #0: c10::Error::Error(c10::SourceLocation, std::string const&) + 0x45 (0x7ff4ba0d7cf5 in /usr/local/pytorch/1.0.1/lib/python3.7/site-packages/torch/lib/libc10.so)
frame #1: THStorage_free + 0xca (0x7ff4bac208ea in /usr/local/pytorch/1.0.1/lib/python3.7/site-packages/torch/lib/libcaffe2.so)
frame #2: <unknown function> + 0x12c63d (0x7ff4e13e863d in /usr/local/pytorch/1.0.1/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #3: <unknown function> + 0x1ae326 (0x7ff4f64b1326 in /usr/local/pytorch/1.0.1/bin/python)
frame #4: <unknown function> + 0x10f13c (0x7ff4f641213c in /usr/local/pytorch/1.0.1/bin/python)
frame #5: <unknown function> + 0xfa978 (0x7ff4f63fd978 in /usr/local/pytorch/1.0.1/bin/python)
frame #6: <unknown function> + 0xfe5e8 (0x7ff4f64015e8 in /usr/local/pytorch/1.0.1/bin/python)
frame #7: <unknown function> + 0x10f8d7 (0x7ff4f64128d7 in /usr/local/pytorch/1.0.1/bin/python)
frame #8: <unknown function> + 0x10f8ed (0x7ff4f64128ed in /usr/local/pytorch/1.0.1/bin/python)
frame #9: <unknown function> + 0x10f8ed (0x7ff4f64128ed in /usr/local/pytorch/1.0.1/bin/python)
frame #10: _PyEval_EvalFrameDefault + 0x5cf0 (0x7ff4f64d8dd0 in /usr/local/pytorch/1.0.1/bin/python)
frame #11: _PyEval_EvalCodeWithName + 0x2f9 (0x7ff4f6418f19 in /usr/local/pytorch/1.0.1/bin/python)
frame #12: PyEval_EvalCodeEx + 0x44 (0x7ff4f6419dd4 in /usr/local/pytorch/1.0.1/bin/python)
frame #13: PyEval_EvalCode + 0x1c (0x7ff4f6419dfc in /usr/local/pytorch/1.0.1/bin/python)
frame #14: <unknown function> + 0x22f9e4 (0x7ff4f65329e4 in /usr/local/pytorch/1.0.1/bin/python)
frame #15: PyRun_FileExFlags + 0xa1 (0x7ff4f653cbd1 in /usr/local/pytorch/1.0.1/bin/python)
frame #16: PyRun_SimpleFileExFlags + 0x1c3 (0x7ff4f653cdc3 in /usr/local/pytorch/1.0.1/bin/python)
frame #17: <unknown function> + 0x23aedb (0x7ff4f653dedb in /usr/local/pytorch/1.0.1/bin/python)
frame #18: _Py_UnixMain + 0x3c (0x7ff4f653dfbc in /usr/local/pytorch/1.0.1/bin/python)
frame #19: __libc_start_main + 0xf5 (0x7ff4f5b183d5 in /lib64/libc.so.6)
frame #20: <unknown function> + 0x1dfed2 (0x7ff4f64e2ed2 in /usr/local/pytorch/1.0.1/bin/python)

**Comment by sndnyang** [comment]
I got the same issue.  Hope my code can help. 

My version is  PyTorch version: 1.0.1.post2
Name: torch
Version: 1.0.1.post2

Name: tensorboard
Version: 1.10.0

Name: tensorboardX
Version: 1.4

```
import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as nfunc
def call_bn(bn, x, update_batch_stats=True):
    if bn.training is False:
        return bn(x)
    elif not update_batch_stats:
        return nfunc.batch_norm(x, None, None, bn.weight, bn.bias, True, bn.momentum, bn.eps)
    else:
        return bn(x)
    
    
class MLP(nn.Module):
    def __init__(self, layer_sizes, affine=False, top_bn=True):
        super(MLP, self).__init__()
        self.input_len = 1 * 28 * 28
        self.fc1 = nn.Linear(self.input_len, 1200)
        self.fc2 = nn.Linear(1200, 1200)
        self.fc3 = nn.Linear(1200, 10)

        self.bn_fc1 = nn.BatchNorm1d(1200, affine=affine)
        self.bn_fc2 = nn.BatchNorm1d(1200, affine=affine)
        self.top_bn = top_bn
        if top_bn:
            self.bn_fc3 = nn.BatchNorm1d(10, affine=affine)

    def forward(self, x, update_batch_stats=True):
        h = nfunc.relu(call_bn(self.bn_fc1, self.fc1(x.view(-1, self.input_len)), update_batch_stats))
        h = nfunc.relu(call_bn(self.bn_fc2, self.fc2(h), update_batch_stats))
        if self.top_bn:
            h = call_bn(self.bn_fc3, self.fc3(h), update_batch_stats)
        else:
            h = self.fc3(h)
        logits = h
        return logits

model = MLP(None)
a = model(torch.zeros(2, 1, 28, 28))  # it works fine

dummy_input = (torch.zeros(2, 1, 28, 28),)
writer.add_graph(MLP(None), dummy_input, True)

```




RuntimeError: r ASSERT FAILED at /pytorch/aten/src/ATen/core/jit_type.h:128, please report a bug to PyTorch. (expect at /pytorch/aten/src/ATen/core/jit_type.h:128)
frame #0: std::function<std::string ()>::operator()() const + 0x11 (0x7fbe567d6021 in /home/xyang2/software/dllib3/lib/python3.6/site-packages/torch/lib/libc10.so)

**Comment by suo** [comment]
@sndnyang Can you confirm that this happens on the latest nightly of PyTorch? I am running your repro against master and it does not trip an assertion; possibly this problem has been fixed

**Comment by sndnyang** [comment]
@suo   I'm trying to use tensorboardX to add_graph  using jupyter notebook.   I tested again and on previous and latest PyTorch with 3 different models. 
torch.__version__  '1.0.1.post2'
torch.__version__  '1.1.0.dev20190422'

```
class MLP(nn.Module):
    def __init__(self, layer_sizes, affine=False, top_bn=True):
        super(MLP, self).__init__()
        self.input_len = 1 * 28 * 28
        self.fc1 = nn.Linear(self.input_len, 1200)
        self.fc2 = nn.Linear(1200, 1200)
        self.fc3 = nn.Linear(1200, 10)

        self.bn_fc1 = nn.BatchNorm1d(1200, affine=affine)
        self.bn_fc2 = nn.BatchNorm1d(1200, affine=affine)
        self.top_bn = top_bn
        if top_bn:
            self.bn_fc3 = nn.BatchNorm1d(10, affine=affine)

    def forward(self, x, update_batch_stats=True):
        h = nfunc.relu(self.fc1(x.view(-1, self.input_len)))
        h = self.fc2(h)
        h = self.bn_fc2(h)
        h = nfunc.relu(h)
        h = self.fc3(h)
        logits = h

        return logits

class MLP2(nn.Module):
    def __init__(self, layer_sizes, affine=False, top_bn=True):
        super(MLP2, self).__init__()
        self.input_len = 1 * 28 * 28
        self.fc1 = nn.Linear(self.input_len, 1200)
        self.fc2 = nn.Linear(1200, 1200)
        self.fc3 = nn.Linear(1200, 10)


    def forward(self, x, update_batch_stats=True):
        h = nfunc.relu(self.fc1(x.view(-1, self.input_len)))
        h = self.fc2(h)
        h = nfunc.relu(h)
        h = self.fc3(h)
        logits = h

        return logits

writer = SummaryWriter()
```

The simple model below works fine for both version

```
dummy_input = (torch.zeros(1, 3),)

class LinearInLinear(nn.Module):
    def __init__(self):
        super(LinearInLinear, self).__init__()
        self.l = nn.Linear(3, 5)

    def forward(self, x):
        return self.l(x)
    
writer.add_graph(LinearInLinear(), dummy_input, verbose=True)
graph(%input : Float(1, 3)
      %1 : Float(5, 3)
      %2 : Float(5)) {
  %3 : Float(1, 5) = onnx::Gemm[alpha=1, beta=1, transB=1](%input, %1, %2), scope: LinearInLinear/Linear[l]
  return (%3);
}
```

For MLP2,  it works fine for version 1.0.1.post2, but it fails in '1.1.0.dev20190422' as below. And if I use  `x.view(2, self.input_len)`  instead of `x.view(-1, self.input_len)`,  it's fine.

```
dummy_input = (torch.zeros(2, 1, 28, 28),)

writer.add_graph(MLP2(None), dummy_input, True)

graph(%0 : Float(2, 1, 28, 28),
      %1 : Float(1200, 784),
      %2 : Float(1200),
      %3 : Float(1200, 1200),
      %4 : Float(1200),
      %5 : Float(10, 1200),
      %6 : Float(10)):
  %7 : Tensor = onnx::Constant[value=  -1  784 [ Variable[CPUType]{2} ]](), scope: MLP2
  %8 : Float(2, 784) = onnx::Reshape(%0, %7), scope: MLP2
  %9 : Float(784!, 1200!) = onnx::Transpose[perm=[1, 0]](%1), scope: MLP2/Linear[fc1]
  %10 : Float(2, 1200) = onnx::Gemm[alpha=1, beta=1](%8, %9, %2), scope: MLP2/Linear[fc1]
  %11 : Float(2, 1200) = onnx::Relu(%10), scope: MLP2
  %12 : Float(1200!, 1200!) = onnx::Transpose[perm=[1, 0]](%3), scope: MLP2/Linear[fc2]
  %13 : Float(2, 1200) = onnx::Gemm[alpha=1, beta=1](%11, %12, %4), scope: MLP2/Linear[fc2]
  %14 : Float(2, 1200) = onnx::Relu(%13), scope: MLP2
  %15 : Float(1200!, 10!) = onnx::Transpose[perm=[1, 0]](%5), scope: MLP2/Linear[fc3]
  %16 : Float(2, 10) = onnx::Gemm[alpha=1, beta=1](%14, %15, %6), scope: MLP2/Linear[fc3]
  return (%16)

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-6-407dfa075711> in <module>
      1 dummy_input = (torch.zeros(2, 1, 28, 28),)
----> 2 writer.add_graph(MLP2(None), dummy_input, True)

~/project/other/DeepLearningCMP/venv/lib/python3.6/site-packages/tensorboardX/writer.py in add_graph(self, model, input_to_model, verbose, **kwargs)
    564                     print('add_graph() only supports PyTorch v0.2.')
    565                     return
--> 566             self.file_writer.add_graph(graph(model, input_to_model, verbose))
    567         else:
    568             # Caffe2 models do not have the 'forward' method

~/project/other/DeepLearningCMP/venv/lib/python3.6/site-packages/tensorboardX/pytorch_graph.py in graph(model, args, verbose, omit_useless_nodes)
    238     if verbose:
    239         print(graph)
--> 240     list_of_nodes, node_stats = parse(graph, args, omit_useless_nodes)
    241     stepstats = RunMetadata(step_stats=StepStats(dev_stats=[DeviceStepStats(device="/device:CPU:0",
    242                                                                             node_stats=node_stats)]))

~/project/other/DeepLearningCMP/venv/lib/python3.6/site-packages/tensorboardX/pytorch_graph.py in parse(graph, args, omit_useless_nodes)
    159 
    160     for node in graph.nodes():
--> 161         nodes_py.append(Node_py_OP(node))
    162 
    163     for node in graph.outputs():  # must place last.

~/project/other/DeepLearningCMP/venv/lib/python3.6/site-packages/tensorboardX/pytorch_graph.py in __init__(self, Node_cpp)
     72 class Node_py_OP(Node_py):
     73     def __init__(self, Node_cpp):
---> 74         super(Node_py_OP, self).__init__(Node_cpp, methods_OP)
     75         self.attributes = str({k: Node_cpp[k] for k in Node_cpp.attributeNames()}).replace("'", ' ')
     76         self.kind = Node_cpp.kind()

~/project/other/DeepLearningCMP/venv/lib/python3.6/site-packages/tensorboardX/pytorch_graph.py in __init__(self, Node_cpp, valid_mothods)
     52                         io_tensorSize_list.append(None)
     53                     else:
---> 54                         io_tensorSize_list.append(n.type().sizes())
     55 
     56                 setattr(self, m, io_uniqueName_list)

RuntimeError: r ASSERT FAILED at /pytorch/aten/src/ATen/core/jit_type.h:142, please report a bug to PyTorch. (expect at /pytorch/aten/src/ATen/core/jit_type.h:142)
frame #0: std::function<std::string ()>::operator()() const + 0x11 (0x7f34c43da441 in /home/xyang2/project/other/DeepLearningCMP/venv/lib/python3.6/site-packages/torch/lib/libc10.so)
frame #1: c10::Error::Error(c10::SourceLocation, std::string const&) + 0x2a (0x7f34c43d9d7a in /home/xyang2/project/other/DeepLearningCMP/venv/lib/python3.6/site-packages/torch/lib/libc10.so)
frame #2: std::shared_ptr<c10::CompleteTensorType> c10::Type::expect<c10::CompleteTensorType>() + 0xa3 (0x7f348d776b83 in /home/xyang2/project/other/DeepLearningCMP/venv/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
frame #3: <unknown function> + 0x41c8eb (0x7f348d78b8eb in /home/xyang2/project/other/DeepLearningCMP/venv/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
frame #4: <unknown function> + 0x13113c (0x7f348d4a013c in /home/xyang2/project/other/DeepLearningCMP/venv/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
<omitting python frames>
```

For MLP model, the error information is the same in two version

```
~/software/dllib3/lib/python3.6/site-packages/torch/onnx/symbolic.py in batch_norm(g, input, weight, bias, running_mean, running_var, training, momentum, eps, cudnn_enabled)
    742         assert len(input_sizes) > 1
    743         weight_value = torch.tensor([1.] * input_sizes[1]).type(
--> 744             'torch.' + input.type().scalarType() + 'Tensor')
    745         weight = g.op("Constant", value_t=weight_value)
    746     if bias is None or bias.node().kind() == "prim::Undefined":

RuntimeError: r ASSERT FAILED at /pytorch/aten/src/ATen/core/jit_type.h:128, please report a bug to PyTorch. (expect at /pytorch/aten/src/ATen/core/jit_type.h:128)
frame #0: std::function<std::string ()>::operator()() const + 0x11 (0x7fbd341db021 in /home/xyang2/software/dllib3/lib/python3.6/site-packages/torch/lib/libc10.so)
frame #1: c10::Error::Error(c10::SourceLocation, std::string const&) + 0x2a (0x7fbd341da8ea in /home/xyang2/software/dllib3/lib/python3.6/site-packages/torch/lib/libc10.so)
frame #2: <unknown function> + 0x378e24 (0x7fbc93820e24 in /home/xyang2/software/dllib3/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
frame #3: <unknown function> + 0x11642e (0x7fbc935be42e in /home/xyang2/software/dllib3/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
<omitting python frames>
frame #26: <unknown function> + 0x2fb9e4 (0x7fbc937a39e4 in /home/xyang2/software/dllib3/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
frame #27: torch::jit::BlockToONNX(torch::jit::Block*, torch::jit::Block*, torch::onnx::OperatorExportTypes, std::unordered_map<torch::jit::Value*, torch::jit::Value*, std::hash<torch::jit::Value*>, std::equal_to<torch::jit::Value*>, std::allocator<std::pair<torch::jit::Value* const, torch::jit::Value*> > >) + 0x4e9 (0x7fbc937ff3c9 in /home/xyang2/software/dllib3/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
frame #28: torch::jit::ToONNX(std::shared_ptr<torch::jit::Graph>&, torch::onnx::OperatorExportTypes) + 0x3d6 (0x7fbc93800546 in /home/xyang2/software/dllib3/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
frame #29: <unknown function> + 0x354519 (0x7fbc937fc519 in /home/xyang2/software/dllib3/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
frame #30: <unknown function> + 0x11642e (0x7fbc935be42e in /home/xyang2/software/dllib3/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
```

I'm not sure if it's the issue of tensorboardX or PyTorch. Thank you!

**Comment by rubencart** [comment]
I get a similar error (let me know if I have to open another issue for this, since it's not the exact same error).

### Issue description
I have a master process with a lot of worker processes (>50) loading models from .pth files. Each worker process adds a perturbation to the parameters of the loaded model, evaluates it on a minibatch, and saves it to a different .pth file. Two worker processes might be reading from the same .pth file concurrently (maybe this is a problem?). They also load & save some tensors with torch.load/save to other files, possibly reading/writing concurrently (I do some error catching + time.sleep for this, which seemed to work fine up until now). The issue seems to occur more frequently when evaluating on smaller minibatches (so more saving/loading, more frequently, so possibly more often concurrently to same files/...).

I do not have or use CUDA or a GPU. These errors occasionally occur (I do not know upon doing what exactly, sorry) and kill the worker process. I was running pytorch 1.0.1 (not a post build), then tried with pytorch 1.0.0, but the issue persisted. It is hard to give a code snippet to reproduce the error since my project requires quite some setup and dependencies... (redis, custom data download,...). If you are interested I can make my repo public and provide you with an explanation on how to run it.

### Error
```
terminate called after throwing an instance of 'c10::Error'
  what():  owning_ptr == NullType::singleton() || owning_ptr->refcount_.load() > 0 ASSERT FAILED at /pytorch/c10/util/intrusive_
ptr.h:341, please report a bug to PyTorch. intrusive_ptr: Can only intrusive_ptr::reclaim() owning pointers that were created us
ing intrusive_ptr::release(). (reclaim at /pytorch/c10/util/intrusive_ptr.h:341)
frame #0: std::function<std::string ()>::operator()() const + 0x11 (0x7f286f7defe1 in /home/rubencartuyvels/miniconda/envs/cp3en
v/lib/python3.7/site-packages/torch/lib/libc10.so)
frame #1: c10::Error::Error(c10::SourceLocation, std::string const&) + 0x2a (0x7f286f7dedfa in /home/rubencartuyvels/miniconda/e
nvs/cp3env/lib/python3.7/site-packages/torch/lib/libc10.so)
frame #2: THStorage_free + 0xca (0x7f28348f8aea in /home/rubencartuyvels/miniconda/envs/cp3env/lib/python3.7/site-packages/torch
/lib/libcaffe2.so)
frame #3: <unknown function> + 0x13063d (0x7f2871dbb63d in /home/rubencartuyvels/miniconda/envs/cp3env/lib/python3.7/site-packag
es/torch/lib/libtorch_python.so)
<omitting python frames>
frame #55: __libc_start_main + 0xf0 (0x7f28760a6830 in /lib/x86_64-linux-gnu/libc.so.6)
```
### System Info
I installed pytorch with pip.

```PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: Could not collect

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] numpy==1.16.2
[pip] torch==1.0.0
[pip] torchvision==0.2.2
[conda] torch                     1.0.0                     <pip>
[conda] torchvision               0.2.2                     <pip>
```

**Comment by AlexFuster** [comment]
I had the same problem and at the end it was a matter of types.
in my case I had a float32 in tensor and
````python
#this fails
out=in_tensor/20 

#this works
out=(in_tensor/20).type(torch.float32)
````

**Comment by eellison** [comment]
Hi @AlexFuster, 

Do you mind posting a complete repro ? This error could be triggered in a variety of ways so it's possible that your root cause was different than the other repros.

**Comment by eellison** [comment]
Also, cc @orionr TensorBoard support

**Comment by orionr** [comment]
Hi @sndnyang and others - instead of tensorboardX, can you use the version in PyTorch? We will be focusing our support on that implementation although @lanpa supports both.

Just change

```
from tensorboardX import SummaryWriter
```

to

```
from torch.utils.tensorboard import SummaryWriter
```

and please see if it reproduces. We're looking to increase our test coverage for add_graph, which does have issues with some model types. Thank you for the data! @lanpa, can you add this failure case to our tests that we're looking to fix? @suo, might need your help debugging this as well.

**Comment by sndnyang** [comment]
@orionr   Hi, sorry, I can't import tensorboard
```
In [1]: from torch.utils.tensorboard import SummaryWriter                                                 
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
~/project/other/DLCodeSnippet/venv/lib/python3.6/site-packages/torch/utils/tensorboard/__init__.py in <module>
      1 try:
----> 2     from tensorboard.summary.writer.record_writer import RecordWriter  # noqa F401
      3 except ImportError:

ModuleNotFoundError: No module named 'tensorboard.summary.writer'

During handling of the above exception, another exception occurred:

ImportError                               Traceback (most recent call last)
<ipython-input-1-2775babaea59> in <module>
----> 1 from torch.utils.tensorboard import SummaryWriter

~/project/other/DLCodeSnippet/venv/lib/python3.6/site-packages/torch/utils/tensorboard/__init__.py in <module>
      2     from tensorboard.summary.writer.record_writer import RecordWriter  # noqa F401
      3 except ImportError:
----> 4     raise ImportError('TensorBoard logging requires TensorBoard with Python summary writer installed. '
      5                       'This should be available in 1.14 or above.')
      6 from .writer import FileWriter, SummaryWriter  # noqa F401

ImportError: TensorBoard logging requires TensorBoard with Python summary writer installed. This should be available in 1.14 or above.

```

**Comment by sndnyang** [comment]
By the way, I have installed tensorboard.

In [2]: import tensorboard                                                                                

In [3]:

**Comment by orionr** [comment]
@sndnyang, please try `pip install tb-nightly` for now (it should get to the release channel soon) to get 1.14 and then try again. Let me know if that works for you.

**Comment by sndnyang** [comment]
> @sndnyang, please try `pip install tb-nightly` for now (it should get to the release channel soon) to get 1.14 and then try again. Let me know if that works for you.

@orionr , Hi,  the same question when I run  ` dummy_input = (torch.zeros(2, 1, 28, 28),); writer.add_graph(MLP(None), dummy_input, True)`

```
~/project/other/DLCodeSnippet/venv/lib/python3.6/site-packages/torch/onnx/symbolic.py in batch_norm(g, input, weight, bias, running_mean, running_var, training, momentum, eps, cudnn_enabled)
    994         weight_value = torch.tensor([1.] * input_sizes[1]).type(
--> 995             'torch.' + input.type().scalarType() + 'Tensor')
    996         weight = g.op("Constant", value_t=weight_value)

RuntimeError: r ASSERT FAILED at /pytorch/aten/src/ATen/core/jit_type.h:142, please report a bug to PyTorch. (expect at /pytorch/aten/src/ATen/core/jit_type.h:142)
frame #0: std::function<std::string ()>::operator()() const + 0x11 (0x7ffaf01c9441 in /home/xyang2/project/other/DLCodeSnippet/venv/lib/python3.6/site-packages/torch/lib/libc10.so)
frame #1: c10::Error::Error(c10::SourceLocation, std::string const&) + 0x2a (0x7ffaf01c8d7a in 
```

and what's more (I'm not sure if it's an issue) after the previous exception.

```
During handling of the above exception, another exception occurred:

NameError                                 Traceback (most recent call last)
<ipython-input-6-0645dc03db09> in <module>
      1 dummy_input = (torch.zeros(2, 1, 28, 28),)
----> 2 writer.add_graph(MLP(None), dummy_input, True)

~/project/other/DLCodeSnippet/venv/lib/python3.6/site-packages/torch/utils/tensorboard/writer.py in add_graph(self, model, input_to_model, verbose, **kwargs)
    532                     print('add_graph() only supports PyTorch v0.2.')
    533                     return
--> 534             self._get_file_writer().add_graph(graph(model, input_to_model, verbose, **kwargs))
    535         else:
    536             # Caffe2 models do not have the 'forward' method

~/project/other/DLCodeSnippet/venv/lib/python3.6/site-packages/torch/utils/tensorboard/_pytorch_graph.py in graph(model, args, verbose, operator_export_type, omit_useless_nodes)
    301         # and we don't want graph visualization to fail in this case. In this
    302         # case we'll log the warning and display the non-optimized graph.
--> 303         logging.warn(ImportError(e))
    304     graph = trace.graph()
    305     if verbose:

NameError: name 'logging' is not defined
```
Thank you so much!

**Comment by orionr** [comment]
Thanks for the feedback, @sndnyang - very helpful. We'll clean this up. @lanpa, FYI.

**Comment by orionr** [comment]
@sndnyang we still have some issues with `add_graph` for this model, but if you install PyTorch nightly it shouldn't fail even though the graph optimization fails. If the MLP model visualized doesn't look good you can try and pass `operator_export_type="RAW"` to add_graph and see if that helps.

Once you've confirmed the fixes we've made in nightly cause add_graph to at least not fail I'll close this issue. We'll track fixing the underling JIT optimization separately.

**Comment by sndnyang** [comment]
@orionr  Yes! It can plot now! Great! 
![image](https://user-images.githubusercontent.com/2310591/57400148-dd1ecc80-71a0-11e9-8ff6-5aab48a3ca0e.png)

**Comment by orionr** [comment]
Excellent!
